x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java

```diff
@@ -0,0 +1,147 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Objects;
+
+/**
+ * A {@link SearchableSnapshotIndexInput} instance corresponds to a single file from a Lucene directory that has been snapshotted. Because
+ * large Lucene file might be split into multiple parts during the snapshot, {@link SearchableSnapshotIndexInput} requires a
+ * {@link FileInfo} object at creation time. This object is used to retrieve the file name and length of the original Lucene file, as well
+ * as all the parts (stored as "blobs" in the repository) that composed the file in the snapshot.
+ *
+ * For example, the following {@link FileInfo}:
+ *  [name: __4vdpz_HFQ8CuKjCERX0o2A, numberOfParts: 2, partSize: 997b, partBytes: 997, metadata: name [_0_Asserting_0.pos], length [1413]
+ *
+ * Indicates that the Lucene file "_0_Asserting_0.pos" has a total length of 1413 and is snapshotted into 2 parts:
+ * - __4vdpz_HFQ8CuKjCERX0o2A.part1 of size 997b
+ * - __4vdpz_HFQ8CuKjCERX0o2A.part2 of size 416b
+ *
+ * {@link SearchableSnapshotIndexInput} maintains a global position that indicates the current position in the Lucene file where the
+ * next read will occur. In the case of a Lucene file snapshotted into multiple parts, this position is used to identify which part must
+ * be read at which position (see {@link #readInternal(byte[], int, int)}. This position is also passed over to cloned and sliced input
+ * along with the {@link FileInfo} so that they can also track their reading position.
+ */
+public class SearchableSnapshotIndexInput extends BufferedIndexInput {
+
+    private final BlobContainer blobContainer;
+    private final FileInfo fileInfo;
+    private final long offset;
+    private final long length;
+
+    private long position;
+    private boolean closed;
+
+    public SearchableSnapshotIndexInput(final BlobContainer blobContainer, final FileInfo fileInfo) {
+        this("SearchableSnapshotIndexInput(" + fileInfo.physicalName() + ")", blobContainer, fileInfo, 0L, 0L, fileInfo.length());
+    }
+
+    private SearchableSnapshotIndexInput(final String resourceDesc, final BlobContainer blobContainer,
+                                         final FileInfo fileInfo, final long position, final long offset, final long length) {
+        super(resourceDesc);
+        this.blobContainer = Objects.requireNonNull(blobContainer);
+        this.fileInfo = Objects.requireNonNull(fileInfo);
+        this.offset = offset;
+        this.length = length;
+        this.position = position;
+        this.closed = false;
+    }
+
+    @Override
+    public long length() {
+        return length;
+    }
+
+    private void ensureOpen() throws IOException {
+        if (closed) {
+            throw new IOException(toString() + " is closed");
+        }
+    }
+
+    @Override
+    protected void readInternal(byte[] b, int offset, int length) throws IOException {
+        ensureOpen();
+        if (fileInfo.numberOfParts() == 1L) {
```

Comment:
Why not reuse the logic from the restore codebase that has a sliced input stream already here instead of building the same thing again? (also see my comment on keeping a reference to a stream open until we're seeking backwards).

Reply:
That's a good suggestion, we should be able to use SlicedInputStream combined with the length parameter. I'll take a look :)